README.md

```diff
@@ -159,15 +159,25 @@ any errors encountered during the request.
 
 ## Body
 
-If the server's response contains a body it should follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
+If the server's response contains a body it MUST follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
 
-Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format since the source of the response
-may not be the GraphQL server but instead some intermediary such as API gateways, firewalls, etc.
+Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format.
+The source of the response may not be the GraphQL server, but instead some intermediary such as API gateways, firewalls, etc.
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
-It is recommended to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation become part of the result.
+
+If the server successfully executed parts of the operation and returns a well-formed response
+that contains a non-null key `data`, the status code SHOULD be `2xx`.
```

Comment:
Still on the fence about this part, i think it depends upon what we can do with the response Content-Type.

When we have an unambigious Content-Type such as `application/graphql+json`, we can safely use any `2xx` or even other status codes for valid responses. That would be my preferred option.

If not, we might have to say the status code MUST be exactly `200` for valid responses.

Reply:
This should be `MUST` IMO.

Reply:
Looking at https://tools.ietf.org/html/rfc7231#section-6.3 I also think it MUST be `200`. 

- `201 Created` would seem only appropriate if the operation created a resource either at the request URL or at a location returned in a `Location` header - that seems weird in a GraphQL context.''

- `202 Accepted` would be appropriate if the response is a partial result, sent while the server is working on remaining parts of the operation. 

  That would actually be an interesting feature, but how is then the next "chunk" of results returned?

- `203 Non-Authoritative Information` seems rather special, and I haven't seen that used so far.

- `204 No Content` disqualifies itself 😄 

- `205 Reset Content` also has no content in the response.

Reply:
`201 Created ` might make sense for some mutations.  Otherwise, `200` is the only status code that seems appropriate for a query.

Reply:
I guess the Content-Type does not necessarily say that the request was successful.

We might want to reserve additional status codes such as `202 Accepted` for features such as deferred queries, where the result might be valid GraphQL - but with some extra bits tacked on.

If a great use case for different status codes comes up, e.g. `201` for some mutations proves to be valuable information, we can open up the spec at a later point to allow it.

As of now, i think GraphQL does quite well by including extra information as part of `data`, `errors` or `extensions`, without leaning on status codes. The simplicity and interoperability with existing tools is another great reason for going `200` only for actual GraphQL responses.

Reply:
I totally agree.  I like the idea of reserving other codes for future potential use.  We will definitely have to think about how to convey a partial response.